File: formula-js/src/QuotedTextResolvedValue.ts
import { ResolvedValue } from "./ResolvedValue";

export class QuotedTextResolvedValue implements ResolvedValue {

  constructor(private readonly value: ResolvedValue,
              private readonly prefix: string,
              private readonly suffix: string) {
  }

  asBoolean(): boolean {
    return this.value.asBoolean();
  }

  asNumber(): number {
    return this.value.asNumber();
  }

  asText(): string {
    return this.value.asText();
  }

  asQuotedText(): string {
    return this.prefix + this.asText() + this.suffix;
  }

  map(fn: (value: ResolvedValue) => ResolvedValue): ResolvedValue {
    return fn(this);
  }

  equals(other: ResolvedValue): boolean {
    return this.value.equals(other);
  }

  hasValue(): boolean {
    return this.value.hasValue();
  }

}

File: formula-java/src/main/java/org/formula/QuotedTextResolvedValue.java
package org.formula;

public class QuotedTextResolvedValue extends ResolvedValue {
    private final ResolvedValue text;
    private final String startQuote;
    private final String endQuote;

    public static QuotedTextResolvedValue of(String text, String startQuote, String endQuote) {
        return of(ResolvedValue.of(text), startQuote, endQuote);
    }

    public static QuotedTextResolvedValue of(ResolvedValue value, String startQuote, String endQuote) {
        return new QuotedTextResolvedValue(value, startQuote, endQuote);
    }

    private QuotedTextResolvedValue(ResolvedValue text, String startQuote, String endQuote) {
        this.text = text;
        this.startQuote = startQuote != null ? startQuote : "";
        this.endQuote = endQuote != null ? endQuote : "";
    }

    @Override
    public String asText() {
        return text.asText();
    }

    public String asQuotedText() {
        return startQuote + text.asText() + endQuote;
    }

    @Override
    public int asNumber() {
        return text.asNumber();
    }

    @Override
    public double asDecimal() {
        return text.asDecimal();
    }

    @Override
    public boolean asBoolean() {
        return text.asBoolean();
    }

    @Override
    public String toString() {
        return asQuotedText();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof ResolvedValue that)) {
            return false;
        }

        return text.equals(that);
    }

    @Override
    public int hashCode() {
        return text.hashCode();
    }
}
